***
***

# 1 - Introduction

Le format du fichier que vous venez de télécharger est un **notebook** au format `ipynb`. 
Les notebooks que nous allons utiliser contiennent :

- des cellules de texte
- des cellules de code Python

Les cellules de texte s'affichent directement, mais il est nécessaire de demander
l'exécution des cellules de code Python pour voir le résultat de l'exécution.

Pour que ces cellules soient exécutées, **un environnement d'exécution est nécessaire**. 
Vous avez le choix entre :

- ouvrir le fichier `ipynb` avec l'**environnement intégré** (*IDE*) de votre choix (*Visual Studio Code*, *PyCharm*, ...).
- l'ouvrir **en ligne** depuis [https://notebook.basthon.fr/](https://notebook.basthon.fr/).

> ⚠️  **Dans le cas où vous utilisez Basthon, attention à sauvegarder régulièrement votre travail dans un fichier pour conserver vos modifications.** ⚠️ 

***

## Exercice 1.1 - Exécution de cellules ⭐



**1)** La cellule ci-dessous est un exemple de code Python exécutable.

Placer le curseur dans la cellule et taper `Ctrl+Entrée`. Le code doit s'exécuter et le message `Hello World!` doit s'afficher juste en dessous.

In [8]:
# Exemple de code Python exécutable
print("Hello World!")

Hello World!


**Explication :**

+ La ligne `# Exemple de code Python exécutable` commence par le caractère `#`.
Elle est donc ignorée et n'est pas exécutée : il s'agit d'un **commentaire**.
+ L'instruction `print(...)` déclenche l'**affichage** de l'expression située 
entre les parenthèses.
+ La valeur affichée est celle de la **chaîne de caractères** délimitée par les 
doubles guillemets `"..."`.

**2)** Dans le cas où le code réalise un calcul, le résultat obtenu sera affiché 
en dessous de la cellule suite à son exécution. 

Le code peut être composé de plusieurs lignes, qui seront toutes exécutées, mais 
seul le résultat de la dernière ligne sera affiché.

In [9]:
1 + 1
2 + 2
print("Hello...")
1234567987654321 * 8765432123456789

Hello...


10821521897576589789666112635269

**Explication :**

- Les résultats des deux premiers calculs ne sont pas affichés
- L'instruction `print(...)` demande explicitement l'**affichage** (`Hello...` apparaît donc dans la sortie de la cellule)
- Le résultat du **dernier calcul** est lui aussi affiché : c'est la **sortie de la dernière instruction** de la cellule.

⚠️ *Ce fonctionnement est **très spécifique au notebook**. On privilégie donc plutôt 
des affichages explicites utilisant `print`, qui sont de toute façon nécessaires
lorsqu'on écrit du code en dehors d'un notebook.*

**3)** Avant de lancer l'exécution de la cellule suivante, essayer de deviner
ce qu'elle va afficher, puis confirmer en lançant l'exécution.

In [10]:
print(1+1)
2+2
print(3+3)
4+4
5+5

2
6


10

<details> 
<summary>Affichage attendu</summary>
<tt>2</tt> <i>(obtenu par le <tt>print(1+1)</tt>)</i><br/>
<tt>6</tt> <i>(obtenu par le <tt>print(3+3)</tt>)</i><br/>
<tt>10</tt> <i>(n'est pas explicitement demandé par un <tt>print</tt>, mais c'est la dernière expression calculée)</i><br/>

Les sommes <tt>2+2</tt> et <tt>4+4</tt> sont calculées mais non affichées.
</details>

**4)** Compléter le code de la cellule ci-dessous de manière à obtenir l'affichage suivant :

```
Bonjour
Au revoir
```

In [11]:
# TODO
print("Bonjour")
print("Au revoir")

Bonjour
Au revoir


**5)** Déterminer les trois derniers chiffres de l'entier suivant :

$$\underbrace{123456789 \times 123456789 \times \ldots \times 123456789}_\textrm{7 répetitions de 123456789}$$

*(Votre calculatrice risque d'avoir un problème... )*

In [13]:
123456789**7

437124189620885610010004822109262358637075660656881926429

<details> 
<summary>Résultat attendu</summary>
Les trois derniers chiffres de <tt>437124189620885610010004822109262358637075660656881926429</tt> sont <tt>429</tt>
</details>

***

## Exercice 1.2 - Erreurs syntaxiques et sémantiques ⭐

**1)** Exécuter le code suivant :

In [15]:
# Exemple d'erreur syntaxique
print("Bonjour")
#print(1/)
print("Au revoir")

Bonjour
Au revoir


**Explication :**

- Une erreur **erreur syntaxique** est détectée *(ici il manque l'opérande de la division `/`)*
- Le code n'est **pas du tout exécuté** : en particulier **la ligne située avant l'erreur syntaxique n'est pas exécutée**.

Des cellules présentant des erreurs syntaxiques peuvent dans certains cas perturber 
la lecture des cellules suivantes.

> **Il est donc important de ne pas conserver des erreurs syntaxiques et de 
s'astreindre à les corriger immédiatement ou à commenter le code fautif.**

> **Avant de poursuivre votre lecture**, commencer par commenter avec un `#` la ligne fautive.

**2)** Exécuter le code suivant :

In [16]:
# Exemple d'erreur sémantique
print("Bonjour")
print(1/0)
print("Au revoir")

Bonjour


ZeroDivisionError: division by zero

**Explication :**

- Le code ne présente pas d'erreur syntaxique donc, contrairement à la question **1**,
le code commence à s'exécuter.
- `Bonjour` est donc bien affiché par le premier `print(...)`.
- Une  **erreur sémantique** est rencontrée *(ici une division par `0`)*
- L'exécution du code s'interrompt alors et la fin du code n'est pas exécuté.

> **⚠️ Important ⚠️** : Vous ferez nécessairement des erreurs, qu'elles soient syntaxiques ou 
sémantiques, c'est parfaitement normal.

Votre travail consiste alors à :

+ **lire** les messages d'erreur donnés par l'interpréteur
+ les **analyser** pour pouvoir les **corriger**.
+ **redemander l'exécution** du code corrigé. 

**3)** Le code suivant essaie d'afficher 3 lignes de texte. 

**a)** L'erreur rencontrée est-elle syntaxique ou sémantique ?

**b)** Corriger la cellule pour obtenir une exécution sans erreur.

In [20]:
print("Une belle journée")
print("Pour corriger du code")
print("Qui fait n'importe quoi")

Une belle journée
Pour corriger du code
Qui fait n'importe quoi


**4)** Le code suivant essaie de calculer le rationnel : $$\frac{80}{3 - \frac{1}{3 \times 9}}$$

**a)** L'erreur rencontrée est-elle syntaxique ou sémantique ?

**b)** Corriger la cellule pour obtenir le résultat correct.

In [23]:
print(80/(3-1/(3*9)))

27.0


<details> 
<summary>Affichage attendu</summary>
L'affichage attendu est <tt>27.0</tt> obtenu en parenthésant correctement l'expression :
<tt>80/(3-1/(3*9))</tt>
</details>

***
***

# 2 - Expressions, affectation, séquence

Dans le langage Python on peut stocker la valeur d'une **expression** dans une
**variable**.

Une **variable** est un *nom* qui identifie un emplacement mémoire dans lequel 
on peut stocker une valeur. 

***

## Exercice 2.1 - Variables et affectations ⭐

**1)** Exécuter le code ci-dessous

In [24]:
toto = 5*(6-3)+9
print(toto)

24


**Explication :**

+ Le résultat du calcul de l'**expression** `5*(6-3)+9` (c'est-à-dire l'**entier** `24`) est stocké dans la variable dont l'identifiant est `toto`.
+ Le contenu de la variable `toto` est affiché grâce à l'instruction `print(...)`.

**2)** **Lire** l'exemple ci-dessous **en anticipant** le résultat attendu, puis l'exécuter.

In [25]:
a = 3            # ligne 1
b = 5            # ligne 2
print(a+b)       # ligne 3
print(a*b)       # ligne 4
c = a * a        # ligne 5
print(c)         # ligne 6
a = 0            # ligne 7
print(c + a + b) # ligne 8

8
15
9
14


**Explication :**
- Les lignes `1` et `2` commencent par définir deux variables `a` (dont la valeur est $3$) et `b` (dont la valeur est $5$).  
- La ligne `3` affiche le résultat de l'expression `a+b` qui est évaluée à $3 + 5 = 8$.
- La ligne `4` affiche ensuite le résultat de l'expression `a*b` qui est évaluée à $3 \times 5 = 15$.
- La ligne `5` définit une nouvelle variable `c`, qui contient `a*a` c'est à dire...$3 \times 3 = 9$, affiché à la ligne `6`.
- À la ligne `7` la variable `a` existe déjà, on vient changer la valeur qu'elle contient, pour remplacer le $3$ par $0$. Les affectations précédentes ne sont pas impactées, ce qui a été calculé précédemment l'a été dans un certain contexte, la modification postérieure de `a` ne vient pas modifier la valeur de `c`.
- À la ligne `8`, la variable `c` a toujours pour valeur $9$, `a` a pour valeur $0$, et `b` n'a jamais changé et a pour valeur $5$. L'affichage obtenu est donc $9+0+5 = 14$.



***

## Exercice 2.2 - Ordre d'exécution ⭐

Le code ci-dessous ne s'exécute pas correctement :

```python
y = y - 1
y = 2 * x
x = x + 3 * y
x = 7
```
**1)** Analyser les raisons pour lesquelles il ne peut pas s'exécuter.

**2)** Le copier dans la cellule ci-dessous et remettre ses lignes dans l'ordre de sorte qu'à la fin $x$ contienne la valeur $46$.

In [27]:
# TODO
x = 7
y = 2 * x
y = y - 1
x = x + 3 * y
print(x)

46


***

## Exercice 2.3 - Nommage des variables ⭐

Le code suivant a pour ambition de calculer l'aire d'un triangle, dont on connait la longueur d'une base et la hauteur correspondante.

```python
Base = 8
hauteur = 3
aire = base_triangle * hauteur__triangle / 2
print(Aire)
```

**1)** Analyser les raisons pour lesquelles il échoue à atteindre son objectif.

**2)** Le copier dans la cellule ci-dessous et le corriger pour qu'il se comporte correctement.

In [30]:
# TODO
base_triangle = 8
hauteur_triangle = 3
aire = base_triangle * hauteur_triangle / 2
print(aire)

12.0


<details> 
<summary>Affichage attendu</summary>
12
</details>

> **⚠️ IMPORTANT ⚠️**

> Il est très important de donner un **nom clair et précis** aux variables.

> Python distingue **les majuscules et les minuscules** : `mavariable`, `Mavariable` et `MAVARIABLE` sont des variables différentes.

***

## Exercice 2.4 - Analyse avec PythonTutor ⭐

**1)** Lire le code suivant et essayer de **prévoir l'affichage** qu'il va produire. Vérifier en exécutant le code.

In [31]:
toto = 6
tata = 2+toto
toto = tata+2
print("toto vaut :",toto)
print("tata vaut :",tata)

toto vaut : 10
tata vaut : 8


**2)** Visualiser l'exécution du code sous [PythonTutor](https://pythontutor.com/visualize.html#code=toto%20%3D%206%0Atata%20%3D%202%2Btoto%0Atoto%20%3D%20tata%2B2%0Aprint%28%22toto%20vaut%20%3A%22%2Ctoto%29%0Aprint%28%22tata%20vaut%20%3A%22%2Ctata%29) 
afin d'observer les valeurs des différentes variables au fur et à mesure de 
l'exécution. 

***

## Exercice 2.5 - Application ⭐

Le code suivant :
```python
S = 1000
print(S)
S = S + 100
print(S)
S = S + 200
print(S)
```
correspond à l'affichage jour par jour de l'évolution d'un compte en banque 
décrite par les opérations suivantes :

- Au départ, J+0, le compte est crédité de 1000€. 
- A J+1, on gagne 100€. 
- A J+2, on gagne 200€. 

**Recopier le code dans la cellule ci-dessous et le modifier** afin de s'adapter 
à l'évolution suivante :

- Au départ, J+0, le compte est initialisé à 0€. 
- A J+1, on gagne 100€. 
- A J+2, on double le solde.
- A J+3, on gagne 200€.
- A J+4, on perd 50€.
- A J+5, on divise le solde par 5.

In [34]:
# TODO
s = 0
print(s)
s = s + 100
print(s)
s = s*2
print(s)
s = s + 200
print(s)
s = s - 50
print(s)
s = s/5
print(s)

0
100
200
400
350
70.0


<details> 
<summary>Affichage attendu</summary>
<pre>0
100
200
400
350
70.0</pre>
</details>

**Remarques :** 

- Si l'on change le solde de départ ou le montant d'une des transactions, on **ne modifie le code qu'à un seul endroit**
- Il faut cependant **relancer l'exécution** de la cellule pour que le résultat soit mis à jour.

***
***

# 3 - Types, opérations

Toutes les expressions, les parties d'expressions ont un **type** et une **valeur**.

**Selon le type** d'une expression, on peut effectuer **différentes opérations**.

***

## Exercice 3.1 - Types numériques `int` et `float` ⭐

**1)** Exécuter le code suivant :

In [ ]:
a = 7
b = 7/2
c = 7//2
print("a",type(a),a)
print("b",type(b),b)
print("c",type(c),c)

*Visualiser l'exécution du code sous [PythonTutor](https://pythontutor.com/visualize.html#code=a%20%3D%207%0Ab%20%3D%207/2%0Ac%20%3D%207//2%0Aprint%28%22a%22%2Ctype%28a%29%2Ca%29%0Aprint%28%22b%22%2Ctype%28b%29%2Cb%29%0Aprint%28%22c%22%2Ctype%28c%29%2Cc%29)*

**Explication :**

+ L'expression `type(a)` a pour valeur le **type** de la variable `a`, ici du type entier (`int`)
+ La variable `b` contient un nombre flottant (`float`) qui vaut `3.5` : l'opération `/` retourne un résultat de type `float`
+ La variable `c` contient un nombre entier (`int`) qui vaut `3` : l'opération `//` retourne le résultat de la division euclidienne (de type `int`).

**2)** Pour chacune des expressions suivantes, prédire son **type** et sa **valeur** :
+ `7+9-3`
+ `7+9-3/1`
+ `7+9/3`
+ `7+9-3//1`
+ `7+9//3`
+ `(7**0.5)**2`
+ `(4**0.5)**2`
+ `4**2`

**3)** Vérifier les types et valeurs en complétant la cellule de code suivante avec les affichages adéquats :

In [35]:
# TODO
a=7+9-3
b=7+9-3/1
c=7+9/3
d=7+9-3//1
e=7+9//3
f=(7**0.5)**2
g=(4**0.5)**2
h=4**2
print("a",type(a),a)
print("b",type(b),b)
print("c",type(c),c)
print("a",type(d),d)
print("b",type(e),e)
print("c",type(f),f)
print("a",type(g),g)
print("b",type(h),h)

a <class 'int'> 13
b <class 'float'> 13.0
c <class 'float'> 10.0
a <class 'int'> 13
b <class 'int'> 10
c <class 'float'> 7.000000000000001
a <class 'float'> 4.0
b <class 'int'> 16


***

## Exercice 3.2 - Celsius et Farenheit ⭐

Étant donnée une température $T_F$ donnée en degrés Farenheit, la température 
$T_C$ en degrés Celsius s'obtient grâce à la formule :

$$T_C = \frac{5}{9} (T_F - 32)$$

En particulier un température $T_F = 212 °F$ équivaut à une température $T_C = 100 °C$

Le code suivant tente de réaliser cette conversion et d'afficher le résultat :

In [36]:
T_F = 212
T_C = 5//9*(T_F-32)
print("T_F =",T_F,"°F")
print("T_C =",T_C,"°C")

T_F = 212 °F
T_C = 0 °C


On remarque que le résultat affiché est incorrect.

**1)** Dans le cadre ci-dessous, recopier le code précédent et le corriger afin d'obtenir le bon résultat :

In [37]:
# TODO
T_F = 212
T_C = 5/9*(T_F-32)
print("T_F =",T_F,"°F")
print("T_C =",T_C,"°C")

T_F = 212 °F
T_C = 100.0 °C


**2)** Si ce n'est pas déjà le cas, modifier le code précédent de telle sorte 
qu'aucune variable, expression ou sous-expression ne soit un nombre flottant.

***

## Exercice 3.3 - Type `str` ⭐

**1)** Exécuter le code suivant :

In [38]:
a = "12345"
b = "678"
c = str(910)
d = a+b+c
e = a*10
print("a",type(a),a)
print("b",type(b),b)
print("c",type(c),c)
print("d",type(d),d)
print("e",type(e),e)

a <class 'str'> 12345
b <class 'str'> 678
c <class 'str'> 910
d <class 'str'> 12345678910
e <class 'str'> 12345123451234512345123451234512345123451234512345


*Visualiser l'exécution du code sous [PythonTutor](https://pythontutor.com/visualize.html#code=a%20%3D%20%2212345%22%0Ab%20%3D%20%22678%22%0Ac%20%3D%20str%28910%29%0Ad%20%3D%20a%2Bb%2Bc%0Ae%20%3D%20a%2A10%0Aprint%28%22a%22%2Ctype%28a%29%2Ca%29%0Aprint%28%22b%22%2Ctype%28b%29%2Cb%29%0Aprint%28%22c%22%2Ctype%28c%29%2Cc%29%0Aprint%28%22d%22%2Ctype%28d%29%2Cd%29%0Aprint%28%22e%22%2Ctype%28e%29%2Ce%29)*

**Explication :**

+ Les variables `a` et `b` ne contiennent pas des entiers mais des **chaînes de caractères** (de type `str`)
+ L'instruction `str(...)` permet de **convertir** une variable numérique en chaîne de caractères. La variable `c` contient donc la chaîne de caractères `"910"`.
+ L'opération `+` entre **deux chaînes de caractères** n'est pas une addition mais une **concaténation**.
+ L'opération `*` entre **une chaîne de caractères et un entier** permet une **duplication** de la chaîne.
+ Les variables `c` et `d` sont aussi des chaînes de caractères.

**2)** En utilisant correctement les opérations `+` et `*` et **sans créer de nouvelle chaîne de caractères** avec des délimiteurs `"..."`, parvenir aux 4 affichages suivants :
```
||||||||||
<->|<->|<->|<->|<->|
|<->|<->|<->|<->|<->|
<->|<->|<->|<->|<->|<->
```

In [48]:
a = "<->"
b = "|"
print(b*10) # solution du premier motif
print((a+b)*5)
print((b+a)*5+b)
print((a+b)*5+a)

||||||||||
<->|<->|<->|<->|<->|
|<->|<->|<->|<->|<->|
<->|<->|<->|<->|<->|<->


**3)** On considère le code suivant :
```python
a = "Résultat : "
b = 13**252
c = a+b
print(c)
```
**a)** Expliquer pourquoi ce code ne s'exécute pas correctement.

**b) Recopier ce code** dans la cellule ci-dessous et ne modifier **que la ligne `c = ...`** de façon que la chaîne `c` contiennent effectivement `"Résultat : "` suivi de la valeur de `b`.

In [50]:
# TODO
a = "Résultat : "
b = 13**252
c = a+str(b)
print(c)

Résultat : 51727892094751095398649095767833497181982816056380387640162410796636130835208984435184411299107836702677278822596685797222633178681811729603809417663817929310185215269500197123173776515352642287418252805892064765268903858153138262382414433914413894109474010788440574633278707316081


**4)** Dans le code ci-dessous, modifier **uniquement la ligne `c = ...`** de 
manière à afficher **sous la forme d'une chaîne de caractères** le calcul de 
l'opération de multiplication de `a` par `b` avec son résultat.

**Exemples :**

- Si `a` et `b` valent respectivement `5` et `6` l'affichage doit être 
**exactement** : `5 x 6 = 30`.
- Si ensuite on modifie **uniquement** les valeurs de `a` et `b` respectivement 
à `7` et `9` l'affichage doit être **exactement** : `7 x 9 = 63`

In [53]:
a = 5
b = 6
c = a*b
print(c)

30


***
***

# 4 - Entrées / Sorties

***

## Exercice 4.1 - `input(...)` ⭐

**1)** On considère le code suivant :

In [54]:
a = input('Première entrée')
b = input('Seconde entrée')
c = a+b
print(c)


Première entréeabc
Seconde entréedef
abcdef


*Visualiser l'exécution du code sous [PythonTutor](https://pythontutor.com/visualize.html#code=a%20%3D%20input%28%27Premi%C3%A8re%20entr%C3%A9e%27%29%0Ab%20%3D%20input%28%27Seconde%20entr%C3%A9e%27%29%0Ac%20%3D%20a%2Bb%0Aprint%28c%29%0A)*

L'exécuter pour les couples d'entrées suivantes :

+ `abc` (première entrée) et `def` (seconde entrée)
+ `10` (première entrée) et `26` (seconde entrée)

**a)** Le résultat est-il celui attendu?

**b)** Que peut-on en déduire pour le type des variables `a` et `b`? 

**c)** Vérifier en affichant le type de `a` et `b`.

**2)** Recopier le code précédent dans la cellule ci-dessous :

In [56]:
# TODO
a = int(input('Première entrée '))
b = int(input('Seconde entrée '))
c = a+b
print(c)


Première entrée 10
Seconde entrée 20
30


Modifier les instructions `... = input(...)` en `... = int(input(...))` puis exécuter le code pour les couples d'entrées suivantes :

+ `abc` (première entrée) et `def` (seconde entrée)
+ `10` (première entrée) et `26` (seconde entrée)

Le résultat est-il celui attendu? Vérifier en affichant le type des variables `a`, `b` et `c`.

**3)** Recopier le code précédent dans la cellule ci-dessous, modifier les instructions `... = int(input(...))` en `... = float(input(...))` puis exécuter le code pour les mêmes couples d'entrées que précédemment :

In [57]:
# TODO
a = float(input('Première entrée '))
b = float(input('Seconde entrée '))
c = a+b
print(c)

Première entrée 10
Seconde entrée 26
36.0


***

## Exercice 4.2 - `print(...)` ⭐

**1)** Exécuter le code suivant :

In [58]:
a = 5
b = 6
print(a,'x',b,'=',a*b)
print(a,'x',b,'=',a*b,sep='')
print(a,'x',b,'=',a*b,sep='~~~')
print(a,end='')
print('x',end='')
print(b,end='')
print('=',end='')
print(a*b,end='')

5 x 6 = 30
5x6=30
5~~~x~~~6~~~=~~~30
5x6=30

*Visualiser l'exécution du code sous [PythonTutor](https://pythontutor.com/visualize.html#code=a%20%3D%205%0Ab%20%3D%206%0Aprint%28a%2C%27x%27%2Cb%2C%27%3D%27%2Ca%2Ab%29%0Aprint%28a%2C%27x%27%2Cb%2C%27%3D%27%2Ca%2Ab%2Csep%3D%27%27%29%0Aprint%28a%2C%27x%27%2Cb%2C%27%3D%27%2Ca%2Ab%2Csep%3D%27~~~%27%29%0Aprint%28a%2Cend%3D%27%27%29%0Aprint%28%27x%27%2Cend%3D%27%27%29%0Aprint%28b%2Cend%3D%27%27%29%0Aprint%28%27%3D%27%2Cend%3D%27%27%29%0Aprint%28a%2Ab%2Cend%3D%27%27%29)*

**Explication :**

+ Les différents paramètres séparés par des virgules dans l'instruction `print(...,...,...)` sont afffichés sur la même ligne, **séparés par des espaces** (comportement par défaut)
+ Le paramètre additionnel `sep=...` permet de modifier ce comportement par défaut en précisant la chaîne de caractères à afficher entre chaque paramètre.
+ Le paramètre additionnel `end=...` permet de modifier le caractère affiché en fin de ligne. Par défaut ce caractère est `'\n'` c'est-à-dire le caractère spécial de **retour à la ligne**. L'ajout du paramètre `end=''` permet de ne pas retourner à la ligne entre deux affichages.

**2)** Le code suivant :

```python
print("Lors des courses")
print("il faudra acheter")
print("des pommes","des poires")
print("des scoubidous")
```

génère l'affichage suivant :

```
Lors des courses
il faudra acheter
des pommes des poires
des scoubidous
```

Recopier ce code dans le cadre ci-dessous et le modifier afin d'obtenir l'affichage suivant :

`Lors des courses, il faudra acheter : des pommes, des poires et des scoubidous.`

**Contraintes :**

+ Le résultat doit être **exact au caractère près** (ponctuation **et** espaces).
+ Les **seules modifications autorisées** sont des **ajouts de paramètres supplémentaires** `end` et `sep` judicieusement choisis.

In [61]:
print("Lors des courses",end=', ')
print("il faudra acheter",end=' : ')
print("des pommes","des poires",end= ' et ')
print("des scoubidous",end = '.')

Lors des courses, il faudra acheter : des pommes des poires et des scoubidous.

***

## Exercice 4.3 - Premiers petits programmes ⭐

**1)** Exécuter le code suivant :

In [62]:
a = int(input('a ='))
b = int(input('b ='))
print("a =",a)
print("b =",b)
print("q =",a//b)
print("r =",a%b)

a =10
b =10
a = 10
b = 10
q = 1
r = 0


*Visualiser l'exécution du code sous [PythonTutor](https://pythontutor.com/visualize.html#code=a%20%3D%20int%28input%28%27a%20%3D%27%29%29%0Ab%20%3D%20int%28input%28%27b%20%3D%27%29%29%0Aprint%28%22a%20%3D%22%2Ca%29%0Aprint%28%22b%20%3D%22%2Cb%29%0Aprint%28%22q%20%3D%22%2Ca//b%29%0Aprint%28%22r%20%3D%22%2Ca%25b%29)*

**2)** En s'inspirant du code précédent, écrire dans la cellule ci-dessous un code qui demande à l'utilisateur trois **nombres entiers** `x`, `y` et `z` et calcule et affiche leur **somme** et leur **produit**.

In [64]:
# TODO
a = int(input('a = '))
b = int(input('b = '))
c = int(input('c = '))
print("S = ",a+b+c)
print("X = ",a*b*c)

a = 7
b = 4
c = 15
S =  26
X =  420


**Indication :** Vérifier que lorsque les entrées sont respectivement `7`, `4` et `15`, la somme obtenue doit être `26` et le produit `420`.

**3)** En s'inspirant du code précédent, écrire dans la cellule ci-dessous un code qui demande à l'utilisateur deux **nombres entiers** `a` et `b` et calcule une valeur approchée `x` de la solution $x$ de l'équation $a x + b = 0$.

**Indication :**  On supposera sans le vérifier que l'utilisateur donne un entier `a` non nul.



In [ ]:
# TODO
a = int(input('a = '))
b = int(input('b = '))
print()

***

## Exercice 4.4 - Lancer de dés 🎲⭐

**1)** Le code suivant permet de générer un entier `de` aléatoirement entre `1` et `6` :

In [ ]:
from random import randint

de = randint(1,6)
print("Lancer de dé :",de)

⚠️ *Sous PythonTutor, le nombre aléatoire généré est toujours le même, donc la visualisation avec cet outil n'est pas pertinente ici.*

**2)** Recopier le code précédent dans la cellule ci-dessous et le modifier afin de :

+ Réaliser deux lancers.
+ Afficher le résultat de chacun des dés.
+ Afficher la somme des valeurs obtenues.

**Exemple d'affichage :**
```
Dés : 3 5
Somme : 8
```

In [74]:
# TODO
from random import randint

nb_lance = int(input("nb face de lancé? "))
de = randint(1,nb_lance)
des = randint(1,nb_lance)
print("Dés :",de, des)
print("Somme :",de+des)

nb face de lancé?3
Dés : 3 1
Somme : 4


**3)** Modifier le code précédent afin de permettre à l'utilisateur de choisir le nombre de faces du dé lancé.

***

## Exercice 4.5 : discriminant d'un trinôme ⭐⭐

En vous inspirant des codes précédents, écrire dans la cellule ci-dessous un code qui :

+ Demande à l'utilisateur trois **nombres entiers** `a`, `b` et `c` (représentant les coefficients $a$, $b$ et $c$ du trinôme $a x^2 + b x + c$)
+ Calcule et affiche le discriminant `delta` de l'équation $a x^2 + b x + c = 0$ (donné par la formule $\Delta = b^2 - 4 a c$).

In [68]:
# TODO
a = int(input("a= "))
b = int(input("b= "))
c = int(input("c= "))


**Indication :** Vérifier que lorsque les entrées `a`, `b` et `c` sont respectivement `3`, `6` et `2`, le discriminant `delta` vaut `12`.

***
***

# 5 - Booléens

Une variable ou une expression de type **booléen** (`bool`) ne prend que deux 
valeurs : `True` ou `False`.

Les **opérations sur les booléens** sont définies par les tables suivantes :

**Opérateur `not`**

`p`     | `not p`
--------|--------
`False` | `True`
`True`  | `False`

**Opérateur `and`**

`p`     | `q`     | `p and q`
--------|---------|----------
`False` | `False` | `False`
`False` | `True`  | `False`
`True`  | `False` | `False`
`True`  | `True`  | `True`

**Opérateur `or`**

`p`     | `q`     | `p or q`
--------|---------|----------
`False` | `False` | `False`
`False` | `True`  | `True`
`True`  | `False` | `True`
`True`  | `True`  | `True`


Les **opérateurs de comparaison** `==`, `!=`, `<`, `<=`, `>` et `>=` permettent
la comparaison de variables et d’expressions (quelque soit leur type : `int`, 
`float`, `str`) et renvoient une valeur booléenne.

***

## Exercice 5.1 - Le type `bool` ⭐

**1)** Exécuter le code suivant :

In [75]:
a = 8
b = 13
c = a<=10
d = b<=10
e = c or d
print("a",type(a),a)
print("b",type(b),b)
print("c",type(c),c)
print("d",type(d),d)
print("e",type(e),e)

a <class 'int'> 8
b <class 'int'> 13
c <class 'bool'> True
d <class 'bool'> False
e <class 'bool'> True


*Visualiser l'exécution du code sous [PythonTutor](https://pythontutor.com/visualize.html#code=a%20%3D%208%0Ab%20%3D%2013%0Ac%20%3D%20a%3C%3D10%0Ad%20%3D%20b%3C%3D10%0Ae%20%3D%20c%20or%20d%0Aprint%28%22a%22%2Ctype%28a%29%2Ca%29%0Aprint%28%22b%22%2Ctype%28b%29%2Cb%29%0Aprint%28%22c%22%2Ctype%28c%29%2Cc%29%0Aprint%28%22d%22%2Ctype%28d%29%2Cd%29%0Aprint%28%22e%22%2Ctype%28e%29%2Ce%29)*

**Explication :**

+ Les variable `a` et `b` contiennent deux nombres entiers (type `int`).
+ La variable `c` est le résultat de la comparaison de `a` avec `10`, et est donc 
un booléen (type `bool`) dont la valeur est `True`.
+ De même la variable `d` est de type `bool` et vaut `False`.
+ La variable `e` est le résultat de la disjonction `or` de `c` et `d` et est 
donc de type `bool`, de valeur `True` (car `c` est `True`).

**2)** Pour chacune des expressions booléennes suivantes, prédire sa **valeur** :
+ `not 0==1`
+ `6>0 and 7<5`
+ `8<2 or 7>1`
+ `((not 7<1) or 5<0) and 7<6`
+ `(not 7<1) or (5<0 and 7<6)`
+ `not 7<1 or 5<0 and 7<6`

**3)** Vérifier ces valeurs en complétant la cellule de code suivante avec les affichages adéquats :

In [78]:
# TODO
print(not 0==1)
print(6>0 and 7<5)
print(8<2 or 7>1)
print(((not 7<1) or 5<0) and 7<6)
print((not 7<1) or (5<0 and 7<6))
print(not 7<1 or 5<0 and 7<6)
print(type(not 0==1))
print(type(6>0 and 7<5))
print(type(8<2 or 7>1))
print(type(((not 7<1) or 5<0) and 7<6))
print(type((not 7<1) or (5<0 and 7<6)))
print(type(not 7<1 or 5<0 and 7<6))


True
False
True
False
True
True
<class 'bool'>
<class 'bool'>
<class 'bool'>
<class 'bool'>
<class 'bool'>
<class 'bool'>


***

## Exercice 5.2 - Nombres mystères ⭐

**Rappel :** Un entier `n` est divisible par un entier `p` (non nul) si et seulement si le reste de la division entière de `n` par `p` est `0`, autrement dit si `n%p==0`.

**Pour chacun des cadres suivants, préciser la valeur entière de `n` de telle sorte que les booléens affichés soient tous égaux à `True`.**

In [80]:
# 1)
n = 6
print(n>0)
print(n<10)
print(n%2==0)
print(n%3==0)

True
True
True
True


In [82]:
# 2)
n = 28
print(20<n<30)
print(n%7==0 and (n-1)%3==0)

True
True


In [84]:
# 3)
n = 24
print(10<n<20 and not n%1==0 or 20<n<30 and n%8==0)

True


***

## Exercice 5.3 - Année bissextile ⭐

*Jusqu'en 1582, le **calendrier julien** prévoyait des années bissextiles 
(de 366 jours) tous les 4 ans. Mais cela causait un décalage par rapport à l'année
solaire qui n'est **pas exactement de 365,25 jours** mais 
**approximativement 365,2425 jours**. Cette année-là, on est donc passé au 
**calendrier grégorien** pour être cohérent avec cette valeur, et la définition
des années bissextiles a évolué. Depuis :*
 
> **Une année est bissextile si elle est divisible par 4 mais pas par 100, ou si elle est divisible par 400.**

Dans le cadre ci-dessous, compléter le contenue de la parenthèse du `print(...)` 
avec une **expression booléenne** de façon à ce que le programme affiche le booléen
`True` lorsque l'année entière entrée par l'utilisateur est bissextile, et `False`
sinon.

In [86]:
annee = int(input('Entrer une année entière'))
print(annee%4 == 0 and annee%100 != 0)


: 

**Tester le programme** avec les années suivantes :

année  | affichage attendu
-------|------------------
`2023` | `False`
`2024` | `True`
`2100` | `False`
`2400` | `True`

***
***

# 6 - Librairie `turtle` 🐢 (Pour aller plus loin)

*** 

## Exercice 6.1 - Utiliser la librairie `turtle` 🐢

On peut être amené à enrichir les instructions disponibles en important des bibliothèques complémentaires.  

`Turtle` est une bibliothèque qui permet de construire des dessins en donnant des ordres à une "tortue" qui se déplace sur une feuille correspondant au plan cartésien classique.

![repere](https://i.imgur.com/FJ33dwN.png)

On départ, la "tortue" est placée en $(0,0)$, au centre de la fenêtre graphique, orientée vers l'axe des abscisses.
La tortue est capable de réaliser un certain nombre d'actions différentes, de changer de couleur, de laisser ou ne pas laisser de trace...
Elle peut se déplacer vers un point dont on connait les coordonnées, ou au contraire utiliser des mouvements "relatifs" (forward, left...)
Les coordonnées sont mesurées en pixel et les angles sont donnés en degrés. Les cercles de rayon positif sont parcours dans le sens inverse des aiguille d'une
montre (sens trigonométrique).

> Attention (sous VS code), pour que turtle déclenche le dessin il faut que le notebook soit sauvegardé (penser à faire CTRL+S) !

Le programme ci-dessous dessine le drapeau Hollandais.


In [91]:
from turtle import *
speed(0)

up()
goto(-150,-150)
down()

left(90)

begin_fill()
color('blue')
forward(100)
right(90)
forward(300)
right(90)
forward(100)
right(90)
forward(300)
right(90)
end_fill()

forward(200)

begin_fill()
color('red')
forward(100)
right(90)
forward(300)
right(90)
forward(100)
right(90)
forward(300)
right(90)
end_fill()
# indispensable pour obtenir un affichage et sortir proprement !
done()
try:
    bye()
except:
    print("bye")


: 

Seriez vous capables de réaliser le drapeau italien ?

![italia](https://upload.wikimedia.org/wikipedia/commons/thumb/0/03/Flag_of_Italy.svg/110px-Flag_of_Italy.svg.png)

In [3]:
from turtle import *
speed(0)

up()
goto(-150,-150)
down()

left(90)

begin_fill()
color('green')
forward(300)
right(90)
forward(150)
right(90)
forward(300)
end_fill()
left(90)
begin_fill()
color('white')
forward(150)
left(90)
forward(300)
end_fill()
right(90)
begin_fill()
color('red')
forward(150)
right(90)
forward(300)
right(90)
forward(150)
right(90)
end_fill()
done()
try:
    bye()
except:
    print("bye")

bye


Et le drapeau du Sénegal ?

![senegal](https://upload.wikimedia.org/wikipedia/commons/thumb/f/fd/Flag_of_Senegal.svg/165px-Flag_of_Senegal.svg.png)

In [1]:
from turtle import *
speed(0)

up()
goto(-150,-150)
down()

from turtle import *
speed(0)

up()
goto(-150,-150)
down()

left(90)

begin_fill()
color('green')
forward(300)
right(90)
forward(150)
right(90)
forward(300)
end_fill()
left(180)
begin_fill()
color('yellow')
forward(300)
right(90)
forward(150)
right(90)
forward(300)
end_fill()
begin_fill()
color('red')
left(90)
forward(150)
left(90)
forward(300)
left(90)
forward(150)
end_fill()
left(90)
forward(125)
right(90)
forward(30)
begin_fill()
color('green')
forward(100)
left(140)
forward(100)
left(140)
forward(100)
left(140)
forward(100)
left(140)
forward(100)
end_fill()
done()
try:
    bye()
except:
    print("bye")
try:
    bye()
except:
    print("bye")

bye
